### PR TITLE
Don't auto-inline const functions

### LINKED ISSUE #53451: `fn` nested in `const fn` causes "Cannot create local mono-item for ..." ICE cross-crate.
```rust
// a.rs
#![feature(const_fn)]

pub const fn foo() -> fn() {
    fn bar() {}
    bar
}
```
```rust
// b.rs
extern crate a;
fn main() {
    a::foo()();
}
```
```
error: internal compiler error: librustc_mir/monomorphize/collector.rs:757:
Cannot create local mono-item for DefId(10/0:4 ~ a[8787]::foo[0]::bar[0])
```

Adding `#[inline]` to `bar`'s definition works around the ICE.
Also, the bug doesn't trigger if `foo` is an `#[inline]` non-`const` `fn`, so it's not just that `foo` is inlined cross-crate, but `const fn`-specific behavior is also involved.

cc @oli-obk @michaelwoerister

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #53451

### CODE REVIEW

File: src/test/ui/consts/auxiliary/const_fn_lib.rs [hunk 1]
```diff
@@ -10,6 +10,12 @@
...
+#![feature(const_fn)]
 #![crate_type="rlib"]
 
-pub const fn foo() -> usize { 22 } //~ ERROR const fn is unstable
+pub const fn foo() -> usize { 22 }
+
+pub const fn bar() -> fn() {
+    fn x() {}
+    x
+}
```
Comment (eddyb) [inline src/test/ui/consts/auxiliary/const_fn_lib.rs]: Missing newline at the end of file.

**Comment by eddyb** [comment]
cc @nikomatsakis @michaelwoerister @alexcrichton

**Comment by michaelwoerister** [comment]
Hm, the approach doesn't seem quite right to me. With the PR we would translate all const fns to machine code, even if they are never used anywhere. That would add code bloat. Shouldn't we be able to handle this like we handle regular `#[inline]` functions (or do they fail too?). We should rather adapt `reachable.rs` to detect that the nested function is reachable.

**Comment by oli-obk** [comment]
We also translate normal functions without `#[inline]`, even if they are unused in the crate that defines them. One can still write `#[inline] const fn foo() {}` after this PR.

If we start creating more complex const fns, they would get codegened multiple times if they are auto-inlined without any way to opt back out of the implicit inlining

**Comment by michaelwoerister** [comment]
> We also translate normal functions without #[inline], even if they are unused in the crate that defines them. 

Only, if they are publicly visible.

> If we start creating more complex const fns, they would get codegened multiple times if they are auto-inlined without any way to opt back out of the implicit inlining.

Yes, that's true. Let's do a perf run.

I still think this PR is more of a work-around than a real solution to the problem. The compiler should be able to handle both cases. That it doesn't seems like a bug.

**Comment by oli-obk** [comment]
> The compiler should be able to handle both cases. That it doesn't seems like a bug.

Yes indeed. I was going to fix the real thing and questioned the motive of auto-inlining const fns while I was doing so. Thus this PR

> I still think this PR is more of a work-around than a real solution to the problem. 

The bug is code asking whether functions have an `inline` attribute instead of asking the function's `Instance` whether it wants to be inlined.

**Comment by oli-obk** [comment]
@rust-timer build 5720c63

**Comment by oli-obk** [comment]
@rust-timer build 5720c63ceb9a7b0029646d8a61763ae27f358b2f

**Comment by oli-obk** [comment]
2% perf improvement on some benchmarks, 1% regression on `hyper-opt`

+14% to -6% max memory usage, pretty wild results: https://perf.rust-lang.org/compare.html?start=289ad6e9922683807d455ca0020dc2a8f7bd1a7b&end=5720c63ceb9a7b0029646d8a61763ae27f358b2f&stat=max-rss

**Comment by michaelwoerister** [comment]
Many of those max-rss benchmarks seem to be very flaky to begin with, especially `deep-vector-check`.

OK, let's do this. Const fns that are only every used in a const context and are not publicly visible should still never be instantiated (because the collector won't find any calls to them in that case). 

@bors r+

**Pull request merged**